File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def bfs(maps):
    dx = (-1, 1, 0, 0)
    dy = (0, 0, -1, 1)
    queue = deque()
    queue.append([0, 0])
    
    n = len(maps)
    m = len(maps[0])
    
    visited = [ [False for _ in range(m)] for _ in range(n)]
    visited[0][0]=True
    
    
    while queue:
        now = queue.popleft()
        
        for i in range(4):
            nx = dx[i] + now[0]
            ny = dy[i] + now[1]
            
            if 0<=nx<n and 0<=ny<m and not visited[nx][ny] and maps[nx][ny] != 0:
                queue.append([nx, ny])
                maps[nx][ny] = maps[now[0]][now[1]]+1
                visited[nx][ny]=True
    return maps[n-1][m-1]
def solution(maps):
    answer = bfs(maps)
    
    return answer if answer != 1 else -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    static final int[] dx = {-1, 1, 0, 0};
    static final int[] dy = {0, 0, -1, 1};

    public int solution(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length;
        if(n==1 && m==1){
            return 1;
        }
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{0, 0});
        
        while (!q.isEmpty()) {
            int[] now = q.poll();
            for (int i = 0; i < 4; i++) {
                int nx = now[0]+dx[i];
                int ny = now[1]+dy[i];
                if(nx>=0&&nx<n&&ny>=0&&ny<m &&maps[nx][ny]==1){
                    maps[nx][ny] = maps[now[0]][now[1]]+1;
                    q.offer(new int[]{nx, ny});
                }
            }
        }

        if(maps[n-1][m-1]==1){
            return -1;
        }
        int answer = maps[n-1][m-1];
        return answer;
    }
}